rfcs/sexp.md

```diff
@@ -0,0 +1,104 @@
+# Add a `Sexp` module to the standard library
+
+I propose to add a `Sexp` module to the standard library.
+The goal is twofold:
+
+1. Provide a common type definition that can be shared
+   among libraries that deal with s-expressions.
+
+2. Promote the use of s-expressions as a lightweight
+   human-readable serialization format.
+
+The module would include at least the following definitions:
+
+```
+type t =
+  | Atom of string
+  | List of t list
+
+val equal: t -> t -> bool
+
+val compare: t -> t -> int
+
+val pp: Format.formatter -> t -> unit
+
+val to_string: t -> string
+
+val of_string: string -> t option
```

Comment:
This signature is going to be an end user usability nightmare. I think the overall idea is a rather bad idea and that this task is best left to third-party libraries. 

But if you really want to proceed on this please make sure to:

1) Report accurate locations and meaningfull error messages on parse errors. 
2) Allow to provide appropriate hooks to attach location information to the generic s-expression representation
3) Make sure to document and make precise the actual parsed s-expression format with a BNF. 

You may want to checkout the [Serialk_sexp](https://erratique.ch/software/serialk/doc/Serialk_sexp/index.html) module which provides all of this (along with a few other nice goodies like sexp query combinators).

Reply:
I'm not sure about this suggestion at all, but possibly we could leave `of_string` with its bad signature and a spec of roughly `(of_string (to_string t) = t)`, and then say that proper parsing is better done by a third-party library.

Reply:
The problem is that people will find it convenient to use in their own programs, especially since it has first class support in the stdlib, and will start to use it for configuration files which will lead to UX disasters.